### PR TITLE
=str #18817 add Java API for GraphStage In/OutHandler

### PR DESCRIPTION
**Pull request by drewhk**
Also added a simple stage in Java and a test for it.

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/Shape.scala [hunk 1]
```diff
@@ -45,7 +45,19 @@ sealed abstract class OutPort { self: Outlet[_] ⇒
  * express the internal structural hierarchy of stream topologies).
  */
 object Inlet {
-  def apply[T](toString: String): Inlet[T] = new Inlet[T](toString)
+  /**
+   * Scala API
+   *
+   * Creates a new Inlet with the given name.
+   */
+  def apply[T](name: String): Inlet[T] = new Inlet[T](toString)
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/Shape.scala]: toString?

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/Shape.scala]: Also, in the scaladoc/javadoc, I think it is important to describe the significance of the name (if any)

File: akka-stream/src/main/scala/akka/stream/Shape.scala [hunk 2]
```diff
@@ -62,7 +74,20 @@ final class Inlet[T] private (override val toString: String) extends InPort {
...
  */
 object Outlet {
-  def apply[T](toString: String): Outlet[T] = new Outlet[T](toString)
+
+  /**
+   * Scala API
+   *
+   * Creates a new Outlet with the given name.
+   */
+  def apply[T](name: String): Outlet[T] = new Outlet[T](toString)
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/Shape.scala]: bug (toString is from object Outlet)

File: akka-stream/src/main/scala/akka/stream/Shape.scala (hunk 1 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/Shape.scala]: Why toString? It is the name of the port. What does toString even mean for a parameter? It if course makes sense in the case class, but I don't think it is meaningful for the user of the API.

File: akka-stream/src/main/scala/akka/stream/Shape.scala (hunk 2 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/Shape.scala]: yep, thans!

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala [hunk 3]
```diff
@@ -351,6 +386,11 @@ private[stream] final class GraphInterpreter(
     } else _Name
 
   /**
+   * INTERNAL API
+   */
+  private[stream] def nonNull = this
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: No return type?

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 4]
```diff
@@ -95,14 +95,15 @@ object GraphStageLogic {
    */
   class ConditionalTerminateInput(predicate: () ⇒ Boolean) extends InHandler {
     override def onPush(): Unit = ()
-    override def onUpstreamFinish(): Unit = if (predicate()) inOwnerStageLogic.completeStage()
+    override def onUpstreamFinish(): Unit =
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: Why isn't the GraphStageLogic a parameter to this method instead?

File: akka-stream/src/main/scala/akka/stream/Shape.scala (hunk 1 above)
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/Shape.scala]: It's object Inlet's toString method you're invoking here

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala (hunk 4 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: for posterity my argument from gitter:

> The point of the current proposal is to enable the common defaults to be shared and save their allocations. Passing in the stage would enable that but user code would still have to close over the stage as well in order to do anything interesting, therefore the cluttered API does not save allocations or object size for them. To my mind the syntax overhead does not pull is weight.

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: @rkuhn The nested classes approach seems to work just fine: https://github.com/akka/akka/pull/18862

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: But you cannot preallocate certain handlers. I.e. that wastes memory, this one does not.

**Comment by 2m** [comment]
LGTM. But this PR should unignore these two Java tests aswell:

https://github.com/akka/akka/commit/162bd9b8e365b5e545bc9f555e079efb611cf20c#diff-3b98156e098f2c6b236ff08f1b9d0598R183
https://github.com/akka/akka/commit/162bd9b8e365b5e545bc9f555e079efb611cf20c#diff-98bf7f3f099ab230ec2d012082452b71R111

**Comment by drewhk** [comment]
Do they work without modification?

**Comment by drewhk** [comment]
@2m Those cannot be unignored as they don't pass. Can you please open a ticket about what needs to be done there? We will do it after M1.

File: akka-stream-tests/src/test/java/akka/stream/stage/JavaIdentityStage.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
+package akka.stream.stage;
```
Comment (ktoso) [inline akka-stream-tests/src/test/java/akka/stream/stage/JavaIdentityStage.java]: license header

File: akka-stream-tests/src/test/java/akka/stream/stage/StageTest.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
+package akka.stream.stage;
```
Comment (ktoso) [inline akka-stream-tests/src/test/java/akka/stream/stage/StageTest.java]: license header?

**Comment by 2m** [comment]
Ok. I'll create a ticket.

File: akka-stream-tests/src/test/java/akka/stream/stage/JavaIdentityStage.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+import akka.stream.Attributes;
+import akka.stream.FlowShape;
+import akka.stream.Inlet;
+import akka.stream.Outlet;
+
+public class JavaIdentityStage<T> extends GraphStage<FlowShape<T, T>> {
+  private Inlet<T> _in = Inlet.of("Identity.in");
```
Comment (ktoso) [inline akka-stream-tests/src/test/java/akka/stream/stage/JavaIdentityStage.java]: here `of` reads weirdly... create would sound better

File: akka-stream-tests/src/test/java/akka/stream/stage/JavaIdentityStage.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        });
+
+        setHandler(out(), new AbstractOutHandler() {
+          @Override
+          public void onPull() {
+            pull(in());
+          }
+        });
+
+      }
```
Comment (ktoso) [inline akka-stream-tests/src/test/java/akka/stream/stage/JavaIdentityStage.java]: looks good

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala [hunk 9]
```diff
@@ -222,6 +222,38 @@ private[stream] object GraphInterpreter {
...
+  /**
+   * INTERNAL API
+   */
+  private val _currentInterpreter = new ThreadLocal[Array[AnyRef]] {
+    /*
+     * Using an Object-array avoids holding on to the GraphInterpreter class
+     * when this accidentally leaks onto threads that are not stopped when this
+     * class should be unloaded.
+     */
+    override def initialValue = new Array(1)
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: Sneaky, I get it - we avoid hitting the thread-local, but just touch the pulled out array after it's pulled out... nice.

File: akka-stream-tests/src/test/java/akka/stream/stage/JavaIdentityStage.java (hunk 7 above)
Comment (drewhk) [inline akka-stream-tests/src/test/java/akka/stream/stage/JavaIdentityStage.java]: Well, it is also quite weird to have `create` and `of` depending which sounds better in a certain context. Let's have one, consistent Java creator name then.

**Comment by ktoso** [comment]
LGTM AFAICS, license header would be good to add though.
Naming bikeshed `of/create` is not urgent.

File: akka-stream-tests/src/test/java/akka/stream/stage/JavaIdentityStage.java (hunk 7 above)
Comment (ktoso) [inline akka-stream-tests/src/test/java/akka/stream/stage/JavaIdentityStage.java]: I prefer "what reads best" in general for Java. APIs like guava have been doing this for years - to make using Java libraries bearable, they need to sound "fluent". Maybe that's a discussion for the entire API though, not this PR :)

**Pull request merged**